File: scripts/verify-build-model.py
#!/usr/bin/env python3
"""验证 Reactor 的默认与 Maven Central effective POM 签名开关。"""

from __future__ import annotations

import subprocess
import sys
import tempfile
import xml.etree.ElementTree as element_tree
import os
from pathlib import Path

NAMESPACE = {"m": "http://maven.apache.org/POM/4.0.0"}
ROOT = Path(__file__).resolve().parents[1]
MVNW = ROOT / "mvnw"
GPG_GROUP = "org.apache.maven.plugins"
GPG_ARTIFACT = "maven-gpg-plugin"


def text(element: element_tree.Element | None) -> str | None:
    return element.text.strip() if element is not None and element.text else None


def java_17_required() -> None:
    result = subprocess.run(["java", "-version"], text=True, capture_output=True, check=False)
    output = result.stderr + result.stdout
    if result.returncode != 0 or 'version "17' not in output:
        raise RuntimeError(f"需要 Java 17，实际输出：{output.strip()}")


def recursive_reactor_poms(pom: Path) -> list[Path]:
    model = element_tree.parse(pom).getroot()
    result = [pom]
    for module in model.findall("m:modules/m:module", NAMESPACE):
        module_pom = pom.parent / (module.text or "") / "pom.xml"
        if not module_pom.is_file():
            raise RuntimeError(f"Reactor module 不存在：{module_pom}")
        result.extend(recursive_reactor_poms(module_pom))
    return result


def effective_pom(
    pom: Path, profile: str | None, output: Path, cache: Path | None
) -> element_tree.Element:
    command = [str(MVNW), "-B", "-N", "-f", str(pom)]
    if cache is not None:
        command.append(f"-Dmaven.repo.local={cache}")
    if profile:
        command.append(f"-P{profile}")
    command.extend(["help:effective-pom", f"-Doutput={output}"])
    result = subprocess.run(command, cwd=ROOT, text=True, capture_output=True, check=False)
    if result.returncode != 0 or not output.is_file():
        raise RuntimeError(
            f"无法生成 {pom.relative_to(ROOT)} 的 {profile or 'default'} effective POM：\n"
            f"{result.stdout}\n{result.stderr}"
        )
    return element_tree.parse(output).getroot()


def bootstrap_reactor_metadata(cache: Path | None) -> None:
    bootstrap_poms = (
        ROOT / "pom.xml",
        ROOT / "mimir-boot-bom" / "pom.xml",
        ROOT / "mimir-boot-parent" / "pom.xml",
    )
    for pom in bootstrap_poms:
        command = [
            str(MVNW),
            "-B",
            "-N",
            "-f",
            str(pom),
            "install",
            "-Dgpg.skip=true",
        ]
        if cache is not None:
            command.append(f"-Dmaven.repo.local={cache}")
        result = subprocess.run(command, cwd=ROOT, text=True, capture_output=True, check=False)
        if result.returncode != 0:
            raise RuntimeError(
                f"无法向隔离缓存安装 {pom.relative_to(ROOT)} 元数据：\n{result.stdout}\n{result.stderr}"
            )


def require_gpg_configuration(model: element_tree.Element, expected_skip: str, label: str) -> None:
    properties = model.findall("m:properties/m:gpg.skip", NAMESPACE)
    values = [text(item) for item in properties]
    if len(properties) != 1 or values != [expected_skip]:
        raise RuntimeError(f"{label}: gpg.skip 必须唯一且为 {expected_skip}，实际为 {values}")

    plugins = []
    for plugin in model.findall(".//m:build/m:plugins/m:plugin", NAMESPACE):
        group_id = text(plugin.find("m:groupId", NAMESPACE)) or GPG_GROUP
        artifact_id = text(plugin.find("m:artifactId", NAMESPACE))
        if group_id == GPG_GROUP and artifact_id == GPG_ARTIFACT:
            plugins.append(plugin)
    if not plugins:
        raise RuntimeError(f"{label}: effective POM 缺少 {GPG_ARTIFACT}")

    for plugin in plugins:
        top_level_skip = text(plugin.find("m:configuration/m:skip", NAMESPACE))
        if top_level_skip != expected_skip:
            raise RuntimeError(f"{label}: GPG plugin 顶层 skip={top_level_skip!r}，期望 {expected_skip}")
        executions = plugin.findall("m:executions/m:execution", NAMESPACE)
        if not executions:
            raise RuntimeError(f"{label}: GPG plugin 缺少 execution")
        for execution in executions:
            execution_skip = text(execution.find("m:configuration/m:skip", NAMESPACE))
            if execution_skip != expected_skip:
                execution_id = text(execution.find("m:id", NAMESPACE)) or "<unnamed>"
                raise RuntimeError(
                    f"{label}: GPG execution {execution_id} skip={execution_skip!r}，期望 {expected_skip}"
                )


def main() -> int:
    java_17_required()
    poms = recursive_reactor_poms(ROOT / "pom.xml")
    if len(set(poms)) != len(poms):
        raise RuntimeError("Reactor POM 枚举出现重复")
    with tempfile.TemporaryDirectory(prefix="mimir-effective-poms-") as temporary_directory:
        temporary_root = Path(temporary_directory)
        cache_setting = os.environ.get("MIMIR_BUILD_MODEL_M2")
        cache = Path(cache_setting) if cache_setting else None
        if cache is not None:
            cache.mkdir(parents=True, exist_ok=True)
        bootstrap_reactor_metadata(cache)
        for index, pom in enumerate(poms):
            for profile, expected in ((None, "true"), ("maven-central", "false")):
                model = effective_pom(
                    pom,
                    profile,
                    temporary_root / f"{index}-{profile or 'default'}.xml",
                    cache,
                )
                require_gpg_configuration(model, expected, f"{pom.relative_to(ROOT)} [{profile or 'default'}]")
    print(f"已验证 {len(poms)} 个 Reactor POM 的 default/maven-central GPG 模型")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except (RuntimeError, subprocess.SubprocessError, element_tree.ParseError) as error:
        print(f"构建模型验证失败：{error}", file=sys.stderr)
        raise SystemExit(1)

File: tools/engineering/src/quality/verify-build-model.mjs
import { spawnSync } from 'node:child_process';
import { mkdir, mkdtemp, readFile, realpath, rm } from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { DOMParser } from '@xmldom/xmldom';

const NAMESPACE = 'http://maven.apache.org/POM/4.0.0';
const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../../../..');
const GPG_GROUP = 'org.apache.maven.plugins';
const GPG_ARTIFACT = 'maven-gpg-plugin';

class BuildModelToolError extends Error {
    constructor(message) {
        super(message);
        this.name = 'BuildModelToolError';
        this.exitCode = 2;
    }
}

function elements(node, ...names) {
    return names.reduce((nodes, name) => nodes.flatMap((parent) =>
        Array.from(parent.childNodes).filter((child) =>
            child.nodeType === 1 && child.namespaceURI === NAMESPACE && child.localName === name)), [node]);
}

function text(node) {
    return node?.textContent.trim() || null;
}

export function parsePom(source) {
    const document = new DOMParser({
        onError: (level, message) => { throw new Error(`XML ${level}: ${message}`); },
    }).parseFromString(source, 'application/xml');
    const root = document.documentElement;
    if (document.doctype || root?.namespaceURI !== NAMESPACE || root.localName !== 'project') {
        throw new Error('必须是无 DTD 的 Maven project XML');
    }
    return root;
}

export function requireGpgConfiguration(model, expectedSkip, label) {
    const values = elements(model, 'properties', 'gpg.skip').map(text);
    if (values.length !== 1 || values[0] !== expectedSkip) {
        throw new Error(`${label}: gpg.skip 必须唯一且为 ${expectedSkip}，实际为 ${JSON.stringify(values)}`);
    }
    // 包括 profile 内的 build/plugins，与原检查范围保持一致；不以 pluginManagement 代替实际插件。
    const builds = Array.from(model.getElementsByTagNameNS(NAMESPACE, 'build'));
    const plugins = builds.flatMap((build) => elements(build, 'plugins', 'plugin')).filter((plugin) =>
        (text(elements(plugin, 'groupId')[0]) || GPG_GROUP) === GPG_GROUP
        && text(elements(plugin, 'artifactId')[0]) === GPG_ARTIFACT);
    if (!plugins.length) throw new Error(`${label}: effective POM 缺少 ${GPG_ARTIFACT}`);
    for (const plugin of plugins) {
        const skip = text(elements(plugin, 'configuration', 'skip')[0]);
        if (skip !== expectedSkip) throw new Error(`${label}: GPG plugin 顶层 skip=${skip}，期望 ${expectedSkip}`);
        const executions = elements(plugin, 'executions', 'execution');
        if (!executions.length) throw new Error(`${label}: GPG plugin 缺少 execution`);
        for (const execution of executions) {
            const executionSkip = text(elements(execution, 'configuration', 'skip')[0]);
            if (executionSkip !== expectedSkip) {
                const id = text(elements(execution, 'id')[0]) || '<unnamed>';
                throw new Error(`${label}: GPG execution ${id} skip=${executionSkip}，期望 ${expectedSkip}`);
            }
        }
    }
}

export async function recursiveReactorPoms(pom, seen = new Set()) {
    const absolute = await realpath(pom);
    if (seen.has(absolute)) throw new Error(`Reactor POM 枚举出现重复：${absolute}`);
    seen.add(absolute);
    const model = parsePom(await readFile(absolute, 'utf8'));
    const result = [absolute];
    for (const module of elements(model, 'modules', 'module')) {
        const name = text(module);
        if (!name) throw new Error(`Reactor module 不能为空：${absolute}`);
        result.push(...await recursiveReactorPoms(path.resolve(path.dirname(absolute), name, 'pom.xml'), seen));
    }
    return result;
}

function command(executable, args, label) {
    let result;
    try {
        result = spawnSync(executable, args, { cwd: ROOT, encoding: 'utf8', maxBuffer: 16 * 1024 * 1024 });
    } catch (error) {
        throw new BuildModelToolError(`${label}：${error.message}`);
    }
    const output = `${result.stdout || ''}\n${result.stderr || ''}`;
    if (result.error || result.status !== 0) {
        throw new BuildModelToolError(`${label}：${result.error?.message || `退出码 ${result.status}，信号 ${result.signal}`}\n${output}`);
    }
    return output;
}

export async function verifyBuildModel() {
    const java = command('java', ['-version'], '无法检查 Java 版本');
    if (!/version "17[."+-]/u.test(java)) throw new BuildModelToolError(`需要 Java 17，实际输出：${java.trim()}`);
    const poms = await recursiveReactorPoms(path.join(ROOT, 'pom.xml'));
    const cacheSetting = process.env.MIMIR_BUILD_MODEL_M2;
    const cache = cacheSetting ? path.resolve(cacheSetting) : null;
    const cacheArgs = cache ? [`-Dmaven.repo.local=${cache}`] : [];
    if (cache) await mkdir(cache, { recursive: true });
    const temporary = await mkdtemp(path.join(os.tmpdir(), 'mimir-effective-poms-'));
    try {
        for (const relative of ['pom.xml', 'mimir-boot-bom/pom.xml', 'mimir-boot-parent/pom.xml']) {
            command(path.join(ROOT, 'mvnw'), ['-B', '-N', '-f', path.join(ROOT, relative),
                'install', '-Dgpg.skip=true', ...cacheArgs], `无法安装 ${relative} 元数据`);
        }
        for (const [index, pom] of poms.entries()) {
            for (const [profile, expected] of [[null, 'true'], ['maven-central', 'false']]) {
                const label = `${path.relative(ROOT, pom)} [${profile || 'default'}]`;
                const output = path.join(temporary, `${index}-${profile || 'default'}.xml`);
                command(path.join(ROOT, 'mvnw'), ['-B', '-N', '-f', pom, ...cacheArgs,
                    ...(profile ? [`-P${profile}`] : []), 'help:effective-pom', `-Doutput=${output}`],
                `无法生成 ${label} 的 effective POM`);
                requireGpgConfiguration(parsePom(await readFile(output, 'utf8')), expected, label);
                process.stdout.write(`已验证 ${label}\n`);
            }
        }
    } finally {
        await rm(temporary, { recursive: true, force: true });
    }
    process.stdout.write(`已验证 ${poms.length} 个 Reactor POM 的 default/maven-central GPG 模型\n`);
}

if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
    try {
        if (process.argv.length > 2) {
            if (process.argv.length !== 3 || process.argv[2] !== '--help') throw new BuildModelToolError('只支持 --help 参数');
            process.stdout.write('用法：bash scripts/engineering.sh build-model\n'
                + '要求 Java 17；可用 MIMIR_BUILD_MODEL_M2 指定 Maven 缓存；会安装本地 Reactor 元数据。\n');
        } else {
            await verifyBuildModel();
        }
    } catch (error) {
        process.stderr.write(`构建模型验证失败：${error.message}\n`);
        process.exitCode = Number.isInteger(error?.exitCode) ? error.exitCode : 1;
    }
}
